### PR TITLE
ratelimiter implementation

### PR DESCRIPTION
**Pull request by TopiSenpai**

### CODE REVIEW

**Comment by TopiSenpai** [comment]
linting is not important here for now :p

File: rest/http_client.go [hunk 1]
```diff
@@ -102,11 +128,17 @@ func (c *HTTPClientImpl) DoWithHeaders(ctx context.Context, compiledRoute *route
 		return NewError(rs, err)
 	}
 
+	// catch rate limit errors here
+	// TODO: we should retry on hitting 429
```
Comment (arynxd) [inline rest/http_client.go:132]: unless the API can incorrectly 429, no we should just wait for retry-after

File: rest/http_client.go [hunk 2]
```diff
@@ -20,21 +21,38 @@ type HTTPClient interface {
...
-	return &HTTPClientImpl{userAgent: userAgent, httpClient: httpClient, logger: logger}
+func NewHTTPClient(logger log.Logger, httpClient *http.Client, rateLimiter rate.RateLimiter, headers http.Header, userAgent string) HTTPClient {
+	if logger == nil {
+		logger = log.Default()
+	}
+	if httpClient == nil {
+		httpClient = http.DefaultClient
+	}
+	if rateLimiter == nil {
+		rateLimiter = rate.NewRateLimiter(logger)
```
Comment (arynxd) [inline rest/http_client.go:39]: would it be worth having a RateLimiterConfig to configure the behaviour of the ratelimiter? eg; handling retries and what should happen on a 429>

**Review by arynxd — CHANGES_REQUESTED** [review]
submitted as request changes so we can discuss this before merging, no changes required yet

**Comment by arynxd** [comment]
should we also have a ratelimited event? so users can provide custom logging impls for ratelimits. this event could be fired when we avoid a ratelimit aswell as when we actually hit 429

File: rest/http_client.go (hunk 2 above)
Comment (TopiSenpai) [reply rest/http_client.go:39]: My thought was putting that into the http client as that also handles the retry itself
But yes we should

File: rest/http_client.go (hunk 1 above)
Comment (TopiSenpai) [reply rest/http_client.go:132]: Yup 
This was meant with the correct delay
Just need to grab the bucket again in a loop over retry count

**Comment by Skye-31** [comment]
> should we also have a ratelimited event? so users can provide custom logging impls for ratelimits. this event could be fired when we avoid a ratelimit aswell as when we actually hit 429

I think a separate event for each would be better, something like RatelimitPrevented and RatelimitReceived

File: rest/http_client.go (hunk 2 above)
Comment (TopiSenpai) [reply rest/http_client.go]: nvm I ended up putting it in the RateLimiter Config haha

**Review by TopiSenpai — COMMENTED** [review]
I'll implement the event later

**Pull request merged**